packages/react-pages/src/node/dynamic-modules/PageStrategy.ts

```diff
@@ -0,0 +1,282 @@
+          delete fileCache[file.path]
+          if (file.pageId) {
+            delete pageCache[file.pageId]
+          }
+        } else if (result) {
+          setPageData((file.pageId = result.pageId), result)
+        }
+      }
+
+      watchers.add(
+        chokidar
+          .watch(globs, {
+            cwd: baseDir,
+            ignored: ['**/node_modules/**/*'],
+          })
+          .on(
+            'all',
+            (type, filePath) =>
+              type !== 'addDir' &&
+              type !== 'unlinkDir' &&
+              onPageUpdate(processFile(type, filePath))
+          )
+      )
+    }
+
+    const helpers: PageHelpers = {
+      extractStaticData,
+      loadPageData: (file) => Promise.resolve(loadPageData(file, helpers)),
+      setPageData,
+      removePageData,
+      watchFiles,
+    }
+
+    let pagesPromise = Promise.resolve(pageCache)
+    findPages(pagesDir, helpers)
+
+    // Ensure file changes are processed before resolving `getPages` call.
+    const onPageUpdate = (update: Promise<void>) => {
+      pagesPromise = Promise.all([pagesPromise, update]).then(() => pageCache)
+      emitPromise()
+    }
+
+    this.getPages = () => pagesPromise
+    this.close = () => watchers.forEach((w) => w.close())
+  }
+}
+
+export interface FindPages {
+  (pagesDir: string, helpers: PageHelpers): void
+}
+
+export interface LoadPageData {
+  (file: File, helpers: PageHelpers): PageData | Promise<PageData>
+}
+
+export interface PageData {
+  /**
+   * The page route path.
+   * User can register multiple page data with same pageId,
+   * as long as they have different keys.
+   * Page data with same pageId will be merged.
+   *
+   * @example '/posts/hello-world'
+   */
+  readonly pageId: string
+  /**
+   * The data key.
+   * If it conflicts with an already-registered data,
+   * error will be thrown.
+   *
+   * @default 'main'
+   */
+  readonly key?: string
+  /**
+   * The path to the runtime data module
+   */
+  readonly dataPath?: string
+
+  readonly staticData?: any
+}
+
+export interface PageHelpers {
+  /**
+   * Read the static data from a file.
+   */
+  readonly extractStaticData: (
+    file: File
+  ) => Promise<{
+    readonly [key: string]: any
+    readonly sourceType: string
+  }>
+  /**
+   * Watch files for changes (not necessarily pages).
+   *
+   * Return page data to have deletion handled automatically.
+   * If no handler is passed, the `loadPageData` helper is used
+   * on every matching file.
+   */
+  readonly watchFiles: WatchFilesHelper
+  /**
+   * Load page data using the default page loader.
+   */
+  readonly loadPageData: (file: File) => Promise<PageData>
+  /**
+   * Upsert page data manually.
+   */
+  readonly setPageData: (pageId: string, pageData: Partial<PageData>) => void
+  /**
+   * Remove page data manually.
+   */
+  readonly removePageData: (pageId: string, dataKey?: string) => void
+}
+
+export class File {
+  pageId: string | null = null
```

Comment:
A file may be associaged with multiple page. We should maintain a map for that.

Reply:
Worth noting that if `setPageData` is called by custom `findPages`, we can't know which file is associated with it, so the page may not be invalidated when it should be.

If we force file handling to be serial instead of parallel, we *can* know which file triggered a `setPageData` call. Should we do that?

Reply:
Another thing: If a page is associated with multiple files, we either need to (a) partially invalidate the page on file change, or (b) process every associated file when any of them change.

Reply:
> Worth noting that if setPageData is called by custom findPages, we can't know which file is associated with it, so the page may not be invalidated when it should be.

When `setPageData` is called with `dataPath` (which is the file path), I think we can know which file is associated with the page.

Reply:
The `dataPath` property is optional, isn't it? A page might only have static data and no loaded data.

Reply:
> If we force file handling to be serial instead of parallel, we can know which file triggered a setPageData call. Should we do that?

Yeah. We should call `processFile` serially (use a queue). It will avoid many subtle bugs.

Reply:
> The dataPath property is optional, isn't it? A page might only have static data and no loaded data.

Yes. In that case, no `file <->page relation` should be added.

Reply:
Here is the mental model in my mind: 
1. User call `helpers.watchFiles` in `findPages`
2. When file changes are observed, process each of them**serially** like this:
  - 'add': `handle` it and get `PageData`. Asssociate `PageData.pageId` with the file in a map(let's call it `pageFileMap`). Schedule an update for `@!virtual-modules/pages`.
  - `change`: `handle` it and get `PageData`. Update `pageFileMap`. Schedule an update for `@!virtual-modules/pages/${pageId}`.
  - 'delete': Find all pages asssociated with it. Delte the dataKey from these pages. (If there is no other data, delete the whole page.) Schedule an update for each of these pages (if it is not deleted). Schedule an update for `@!virtual-modules/pages` if some pages is deleted.

Reply:
@csr632 That's exactly how I implemented it while you were typing that comment :)

https://github.com/vitejs/vite-plugin-react-pages/pull/9/commits/89f6380e9538f3f7e7c4b0ae3214d0d4cd563581

Reply:
One quick fix before you merge: I need to make sure `unlink` events are always processed.